pandas/core/internals/array_manager.py

```diff
@@ -1085,15 +1091,24 @@ def unstack(self, unstacker, fill_value) -> ArrayManager:
         unstacked : BlockManager
         """
         indexer, _ = unstacker._indexer_and_to_sort
-        new_indexer = np.full(unstacker.mask.shape, -1)
-        new_indexer[unstacker.mask] = indexer
+        if unstacker.mask.all():
+            new_indexer = indexer
+            allow_fill = False
+        else:
+            new_indexer = np.full(unstacker.mask.shape, -1)
+            new_indexer[unstacker.mask] = indexer
+            allow_fill = True
         new_indexer2D = new_indexer.reshape(*unstacker.full_shape)
+        new_indexer2D = ensure_int64(new_indexer2D)
```

Comment:
can this chunk of code be made to match Block._unstack (or vice-versa) more closely?

e.g. on L1094 you check unstacker.mask.all() which I think is related to the comment in Block._unstack `# TODO: in all tests we have mask.all(); can we rely on that?` but bc `mask` is defined there differently it is not obvious

Reply:
Assuming you are referring to the EA-specific `Block.unstack` (as the other one has even less in common): that's for a single column while this is for multiple columns at once, and needs to deal with creating blocks, with placement, etc. So personally I don't see much possibilities to share code. 

> e.g. on L1094 you check unstacker.mask.all() which I think is related to the comment in Block._unstack

As far as I understand the Block version, I don't think that's related. For Block/BlockManager.unstack, that `mask` value somehow indicates which of the columns need to end up in the result (I don't really understand that code, though, as in the ArrayManager I am simply using the result of `unstacker.get_new_columns` without any filtering of that afterwards), while here I only check the `mask` to potentially use a faster `take` for performance-reasons, but that doesn't influence the end result.

Reply:
> I don't think that's related

OK.

>  (I don't really understand that code, though

at some point we should make sure that one of us (active maintainers, not just the two of us) have a handle on this.